### ISSUE TITLE
Override unittest.TextTestResult when running tests with a custom runner

### ISSUE THREAD

**Issue by felipearcaro** [craigahobbs/unittest-parallel#23]
#19 follow up.

I tested version `1.7.0` on TeamCity and although I was able to run unit tests using the TeamCity custom runner, it seems like there is more work to be done to let the build script interact with the TeamCity server through [service messages](https://www.jetbrains.com/help/teamcity/service-messages.html?Build%2BScript%2BInteraction%2Bwith%2BTeamCity).

I did some research and here's what I found out: 

[TeamcityTestResult](https://github.com/JetBrains/teamcity-messages/blob/9462304bba54c2b37eb373ea715cd73b79bf056b/teamcity/unittestpy.py#L18) extends `unittest.TestResult` and integrates with TeamCity by sending formatted messages using [TeamcityServiceMessages](https://github.com/JetBrains/teamcity-messages/blob/9462304bba54c2b37eb373ea715cd73b79bf056b/teamcity/messages.py#L46). It overrides methods like `addSuccess`, `addError` and `addFailure` to send [service messages](https://www.jetbrains.com/help/teamcity/service-messages.html?Build%2BScript%2BInteraction%2Bwith%2BTeamCity) to TeamCity:

```python
def addSuccess(self, test):
        super(TeamcityTestResult, self).addSuccess(test)
```

In contrast, [ParallelTextTestResult](https://github.com/craigahobbs/unittest-parallel/blob/61405379c6a8aedb5aac8c4b9054b6654ef61582/src/unittest_parallel/main.py#L321) customizes the output by printing messages directly to the stream:

```python
def _add_helper(self, test, dots_message, show_all_message):
        if self.showAll:
            self.stream.writeln(f'{self.getDescription(test)} ... {show_all_message}')
        elif self.dots:
            self.stream.write(dots_message)
        self.stream.flush()

def addSuccess(self, test):
        super(unittest.TextTestResult, self).addSuccess(test)
        self._add_helper(test, '.', 'ok')
```

Do you think there's a simple way to get this done or it would require to fork the project and make it more TeamCity-specific? I am happy to implement it if you think it's viable.

**Comment by craigahobbs** [comment]
I agree - overriding the runner's `resultclass` is also required. This can be done in the same fashion that the runner is overridden. I'll push a branch shortly for you to try out. Out of curiosity, what functionality is in your runner beyond `TextTestRunner`? Could this be accomplished just by overriding the `resultclass`?

**Comment by craigahobbs** [comment]
The branch is up:

https://github.com/craigahobbs/unittest-parallel/tree/feature/custom-result

There is now a `--result` argument. As you've observed, the unittest_parallel `resultclass` ignores the runner's stream. This is because unittest_parallel passes a `StringIO` object to eat `TextTestRunner`'s output, which creates a mess when running in parallel. Now, when using a custom runner class or a custom result class, `None` is passed as the runner's stream as is done with plain-old `unittest.main`.

Also, I've removed the `-r` form of the `--runner` argument - that way I don't have to figure out a single-character argument form for the `--result` argument. :)

**Comment by felipearcaro** [comment]
Hey @craigahobbs - I ran my tests running `unittest-parallel -s . -t . -p '*test*.py' --runner teamcity.unittestpy.TeamcityTestRunner --result teamcity.unittestpy.TeamcityTestRunner` and got the following error:

```
##teamcity[testCount timestamp='2024-12-16T13:08:23.282' count='15']
##teamcity[testCount timestamp='2024-12-16T13:08:23.316' count='8']
##teamcity[testCount timestamp='2024-12-16T13:08:23.359' count='19']
##teamcity[testCount timestamp='2024-12-16T13:08:23.410' count='10']
##teamcity[testCount timestamp='2024-12-16T13:08:23.457' count='10']
##teamcity[testCount timestamp='2024-12-16T13:08:23.479' count='8']
##teamcity[testCount timestamp='2024-12-16T13:08:23.515' count='11']
##teamcity[testCount timestamp='2024-12-16T13:08:23.525' count='23']
##teamcity[testCount timestamp='2024-12-16T13:08:23.526' count='20']
##teamcity[testCount timestamp='2024-12-16T13:08:23.542' count='30']
##teamcity[testCount timestamp='2024-12-16T13:08:23.570' count='24']
##teamcity[testCount timestamp='2024-12-16T13:08:23.571' count='7']
##teamcity[testCount timestamp='2024-12-16T13:08:23.573' count='21']
##teamcity[testCount timestamp='2024-12-16T13:08:23.576' count='10']
##teamcity[testCount timestamp='2024-12-16T13:08:23.618' count='13']
##teamcity[testCount timestamp='2024-12-16T13:08:23.627' count='17']
##teamcity[testCount timestamp='2024-12-16T13:08:23.632' count='14']
##teamcity[testCount timestamp='2024-12-16T13:08:23.634' count='13']
##teamcity[testCount timestamp='2024-12-16T13:08:23.634' count='28']
##teamcity[testCount timestamp='2024-12-16T13:08:23.640' count='16']
##teamcity[testCount timestamp='2024-12-16T13:08:23.674' count='13']
##teamcity[testCount timestamp='2024-12-16T13:08:23.753' count='11']
multiprocessing.pool.RemoteTraceback:
"""
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/usr/local/lib/python3.10/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/usr/local/lib/python3.10/site-packages/unittest_parallel/main.py", line 306, in run_tests
    result = runner.run(test_suite)
  File "/usr/local/lib/python3.10/site-packages/teamcity/unittestpy.py", line 310, in run
    return super(TeamcityTestRunner, self).run(test)
  File "/usr/local/lib/python3.10/unittest/runner.py", line 184, in run
    test(result)
  File "/usr/local/lib/python3.10/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/usr/local/lib/python3.10/unittest/suite.py", line 108, in run
    if result.shouldStop:
AttributeError: 'TeamcityTestRunner' object has no attribute 'shouldStop'
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/bin/unittest-parallel", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.10/site-packages/unittest_parallel/main.py", line 132, in main
    results = pool.map(test_manager.run_tests, test_suites)
  File "/usr/local/lib/python3.10/multiprocessing/pool.py", line 367, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/local/lib/python3.10/multiprocessing/pool.py", line 774, in get
    raise self._value
AttributeError: 'TeamcityTestRunner' object has no attribute 'shouldStop'
root@c081b120241f:/opt/airflow#
```

If you want to try to replicate it, you just need to install the `teamcity-messages` package.

**Comment by craigahobbs** [comment]
The result class must have the same interface as `unittest.TextTestResult`. Your command line specifies the result class of `teamcity.unittestpy.TeamcityTestRunner` which at least missing the `shouldStop` method.

For reference, here's unittest-parallel's result implementation, which inherits its `shouldStop` method from `unittest.TextTestResult`:

https://github.com/craigahobbs/unittest-parallel/blob/61405379c6a8aedb5aac8c4b9054b6654ef61582/src/unittest_parallel/main.py#L321

**Comment by felipearcaro** [comment]
You're right - I ran `unittest-parallel -s . -t . -p '*test*.py' --runner teamcity.unittestpy.TeamcityTestRunner --result teamcity.unittestpy.TeamcityTestResult` (different `result` value) this time and it seems to have worked perfectly. Feel free to merge into to `main`, and thanks for working on it!

**Comment by craigahobbs** [comment]
I'm glad it worked!! :) Published unittest-parallel 1.7.1.
